### PR TITLE
Start a Maestro file (.mae) writer.

### PR DESCRIPTION
**Pull request by ricrogz**
This adds an initial implementation of the `MaeWriter` class, which allows exporting mols to Maestro files (.mae). It is based on the `MolWriter` class, and works in a way similar to the other existing "Writer" classes. The implementation is still very basic, but we hope to improve it soon.

Some things to take into account when using this class:

- The writer does not yet support stereochemistry: Maestro files use absolute stereochemical labels (R/S/E/Z), which require knowledge of the CIP ranks of the atoms around the stereo feature. The new stereo algorithm in RDKit does not provide this information, and while the CIP labeler can calculate the labels, it doesn't currently provide CIP ranks information.

- Kekulization is always required, as there is no concept of aromatic bonds in the Maestro format. Molecules that fail kekulization will not be exported.

- Only single, double, triple, zero and dative bonds are supported in Maestro files.

- Due to the underlying maeparser infrastructure, and same as with `MaeMolSupplier`, `MaeWriter` always requires ownership of the steam it writes to.

- Some other things that neither `MaeMolSupplier` nor `MaeWriter` support are: dummy atoms, isotopes, queries, enhanced stereo, substance groups. There's probably more than these, but we hope to at least add some of them soon.

- The usage of the `setProps()` method is a bit particular, since Maestro files support properties for molecules, atoms and bonds. By default, `propNames` is empty, and `MaeWriter` will export all properties. If `propNames` is not empty, `MaeWriter` will export the required properties (structure title, atom x/y/z coordinates, atomic number, charge, color in Maestro, pdb residue information, bond from/to and bond order) plus any mol/atom/bond property that matches any of the names in `propNames`. Entries in `propNames` that don't match any property will be ignored. 

- Only boolean, integer, real and string properties can be exported.

- Properties in Maestro files must follow a specific formatting rule. To honor this formatting, exported properties will be prefixed with "X_rdk_", where "X" can be 'b', 'i', 'r' or 's', depending on the value type of the property.

- For a structure to be exported, it must have at least one atom, although it is allowed not to have any bonds.

This PR also adds a couple of changes in `MaeMolSupplier` we noticed while adding the tests: a fix in the handling of duplicate bonds in Maestro files, and an update to strip the "X_rdk_" prefixes from properties imported from .mae files.

Pinging @cdvonbargen and @d-b-w here too so they can comment too.

### CODE REVIEW

File: Code/GraphMol/FileParsers/MolWriters.h [hunk 1]
```diff
@@ -333,6 +343,71 @@ class RDKIT_FILEPARSERS_EXPORT PDBWriter : public MolWriter {
...
+   *     have the concept of an aromatic bond.
+   *  - Ownership of the output stream is mandatory, since it needs
+   *     to be managed though a shared_ptr, as this is what maeparser
+   *     writer takes.
+   ***************************************************************************************/
+ public:
+  /*!
+    \param fileName       : filename to write to (stdout is *not* supported)
+   */
+  MaeWriter(const std::string &fileName);
```
Comment (cdvonbargen) [inline Code/GraphMol/FileParsers/MolWriters.h:375]: Can I annoyingly ask for 
```
static std::string getText(const ROMol &mol);
```

Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:375]: I think you mean `MolToMaeBlock()`

**Comment by greglandrum** [comment]
The changes to the property names in the mae parser should be documented in the `Backwards Incompatible Changes` part of the release notes.

**Comment by greglandrum** [comment]
> * The writer does not yet support stereochemistry: Maestro files use absolute stereochemical labels (R/S/E/Z), which require knowledge of the CIP ranks of the atoms around the stereo feature. The new stereo algorithm in RDKit does not provide this information, and while the CIP labeler can calculate the labels, it doesn't currently provide CIP ranks information.

Does the mae format require both the absolute labels (which we can generate) *and* the ranks (which we don't currently)?

If we really need the ranks: do we need them for all atoms or just the ones around chiral centers/bonds?

**Review by greglandrum — CHANGES_REQUESTED** [review]
Generally looks good, but I've made some suggestions and asked some questions.
Also: this needs some tests for the python wrapper.

**Comment by ricrogz** [comment]
> Does the mae format require both the absolute labels (which we can generate) _and_ the ranks (which we don't currently)?
> 
> If we really need the ranks: do we need them for all atoms or just the ones around chiral centers/bonds?

Maestro stereo specifications are properties formatted like:
```
s_st_Chirality_1 = 2_S_3_4_1_18
s_st_EZ_1 = 7_1_2_3_E 
```
For chiral atoms, the first number is the index of the chiral atom, then comes the chiral label, and then the CIP rank sorted list of neighboring atom indexes. For stereo bonds, we have the CIP ranked list of neighboring atoms, and then the label.

So yes, we need the label and ranks of the atoms around the stereo center.

File: Code/GraphMol/FileParsers/MolWriters.h [hunk 2]
```diff
@@ -14,10 +14,20 @@
 
 #include <RDGeneral/types.h>
 
-#include <string>
 #include <iostream>
+#include <memory>
+#include <string>
```
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:19]: memory is required because of the shared pointers in the `MolWriter` class. Strings were already there, and the header was being pulled in by the ROMol.h include. I just made it implicit so that it doesn't depend on another include.

File: Code/GraphMol/FileParsers/MolWriters.h [hunk 3]
```diff
@@ -333,6 +343,71 @@ class RDKIT_FILEPARSERS_EXPORT PDBWriter : public MolWriter {
...
+
+  //! \brief flush the ostream
+  void flush() override;
+  //! \brief close our stream (the writer cannot be used again)
+  void close();
+
+  //! \brief get the number of molecules written so far
+  unsigned int numMols() const override { return d_molid; }
+
+ protected:
```
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:398]: Ah, I made these protected while I was working on the Python wrappers, as these need access to the default constructor and `dp_ostream`, and from some time I also needed a couple other methods I removed later on. And once the wrappers were working, I forgot to make the other members private.

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:375]: Oh. For some reason, I was convinced that `MolWriter` inherited from `boost::noncopyable`, but now I see it doesn't. I'll try to add that.

File: Code/GraphMol/FileParsers/MolWriters.h [hunk 4]
```diff
@@ -333,6 +343,71 @@ class RDKIT_FILEPARSERS_EXPORT PDBWriter : public MolWriter {
...
+   *     to be managed though a shared_ptr, as this is what maeparser
+   *     writer takes.
+   ***************************************************************************************/
+ public:
+  /*!
+    \param fileName       : filename to write to (stdout is *not* supported)
+   */
+  MaeWriter(const std::string &fileName);
+
+  MaeWriter(std::ostream *outStream);
```
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:377]: Done.

File: Code/GraphMol/FileParsers/MolWriters.h [hunk 5]
```diff
@@ -333,6 +343,71 @@ class RDKIT_FILEPARSERS_EXPORT PDBWriter : public MolWriter {
...
+
+  //! \brief get the number of molecules written so far
+  unsigned int numMols() const override { return d_molid; }
+
+ protected:
+  MaeWriter() = default;  // used in the Python wrapper
+
+  void open();
+
+  std::shared_ptr<schrodinger::mae::Writer> dp_writer = nullptr;
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/MolWriters.h:403]: Are you sure about that? Can't you just forward declare the `schrodinger::mae::Writer` class?

**Comment by greglandrum** [comment]
> > Does the mae format require both the absolute labels (which we can generate) _and_ the ranks (which we don't currently)?
> > If we really need the ranks: do we need them for all atoms or just the ones around chiral centers/bonds?
> 
> Maestro stereo specifications are properties formatted like:
> 
> ```
> s_st_Chirality_1 = 2_S_3_4_1_18
> s_st_EZ_1 = 7_1_2_3_E 
> ```
> 
> For chiral atoms, the first number is the index of the chiral atom, then comes the chiral label, and then the CIP rank sorted list of neighboring atom indexes. For stereo bonds, we have the CIP ranked list of neighboring atoms, and then the label.
> 
> So yes, we need the label and ranks of the atoms around the stereo center.

Is it important that the ranks be "CIP ranks" and globally correct (hard), or would it be ok if we just had the relative atom ranks for a particular atom correct (easy)?

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/MolWriters.h:375]: > Oh. For some reason, I was convinced that `MolWriter` inherited from `boost::noncopyable`, but now I see it doesn't. I'll try to add that.

It should. That's actually a better solution than doing it just for this class

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:403]: Yes, I'm sure. `unique_ptr` needs access to the destructor, apparently.

File: Code/GraphMol/FileParsers/MaeWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+
+  // Required properties
+  setPropertyValue(atomBlock, mae::ATOM_X_COORD, numAtoms, idx, coordinates.x);
+  setPropertyValue(atomBlock, mae::ATOM_Y_COORD, numAtoms, idx, coordinates.y);
+  setPropertyValue(atomBlock, mae::ATOM_Z_COORD, numAtoms, idx, coordinates.z);
+
+  auto atomic_num = static_cast<int>(atom.getAtomicNum());
+  setPropertyValue(atomBlock, mae::ATOM_ATOMIC_NUM, numAtoms, idx, atomic_num);
+
+  // Default heavy atoms to be drawn in grey in Maestro, and H atoms in white.
```
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeWriter.cpp:286]: > Is it important that the ranks be "CIP ranks" and globally correct (hard), or would it be ok if we just had the relative atom ranks for a particular atom correct (easy)?

The neighboring atoms must be correctly sorted according to their CIP ranks.

**Comment by ricrogz** [comment]
> Is it important that the ranks be "CIP ranks" and globally correct (hard), or would it be ok if we just had the relative atom ranks for a particular atom correct (easy)?

The neighboring atoms must be correctly sorted according to their CIP ranks.

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 1 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:375]: That's a separate function, btw. I'll implement that in a separate PR, this one is already big enough.

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h]: nvm, I think I figured it out.

File: Code/GraphMol/FileParsers/MolWriters.h (hunk 1 above)
Comment (cdvonbargen) [reply Code/GraphMol/FileParsers/MolWriters.h:379]: Oh I guess I was looking at https://github.com/rdkit/rdkit/blob/27f711a6585e17c0c2882f6d7ea054a1f7a860eb/Code/GraphMol/FileParsers/MolWriters.h#L149 which would do the same as MolToMaeBlock? Or is there mae header stuff that MaeWriter::getText would return that MolToMaeBlock wouldn't?

Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolWriters.h:379]: Yeah, all the work done there is this:
https://github.com/rdkit/rdkit/blob/4d4b92765197df1e60893858cc68f0eeb62adf0e/Code/GraphMol/FileParsers/SDWriter.cpp#L159-L164

... which is basically calling the free function `_MolToSDStream`, which in turn calls `MolToMolBlock`, and then adds some properties to the output:
https://github.com/rdkit/rdkit/blob/4d4b92765197df1e60893858cc68f0eeb62adf0e/Code/GraphMol/FileParsers/SDWriter.cpp#L113-L123

File: Code/GraphMol/FileParsers/PDBWriter.cpp [hunk 7]
```diff
@@ -7,24 +7,28 @@
...
+#include <iostream>
+#include <map>
 #include <string>
 #include <vector>
-#include <map>
-#include <iostream>
-#include <fstream>
-#include <algorithm>
+
+#ifdef RDK_BUILD_MAEPARSER_SUPPORT
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/PDBWriter.cpp:17]: Doesn't seem like the PDBWriter should need to include stuff from maeparser

File: Code/GraphMol/FileParsers/SDWriter.cpp [hunk 8]
```diff
@@ -7,19 +7,25 @@
...
-#include "FileParsers.h"
 
 #include <fstream>
 #include <iostream>
 #include <sstream>
 #include <string>
+
 #include <boost/any.hpp>
 
+#ifdef RDK_BUILD_MAEPARSER_SUPPORT
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/SDWriter.cpp:18]: Doesn't seem like the SDWriter should need to include stuff from maeparser

File: Code/GraphMol/FileParsers/SmilesWriter.cpp [hunk 9]
```diff
@@ -8,18 +8,23 @@
 //  which is included in the file license.txt, found at the root
 //  of the RDKit source tree.
 //
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#ifdef RDK_BUILD_MAEPARSER_SUPPORT
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/SmilesWriter.cpp:16]: Doesn't seem like the SmilesWriter should need to include stuff from maeparser

File: Code/GraphMol/FileParsers/TDTWriter.cpp [hunk 10]
```diff
@@ -8,21 +8,26 @@
...
-
 #include <fstream>
 #include <iostream>
 #include <iomanip>
 #include <sstream>
 #include <string>
+
 #include <boost/algorithm/string.hpp>
 
+#ifdef RDK_BUILD_MAEPARSER_SUPPORT
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/TDTWriter.cpp:19]: Doesn't seem like the TDTWriter should need to include stuff from maeparser

**Review by greglandrum — CHANGES_REQUESTED** [review]
Aside from what looks like a couple of changes which look like the result of an automated transformation applied too broadly, this looks good.
Once those are cleared up I'm 👍 on merging this.

File: Code/GraphMol/FileParsers/PDBWriter.cpp (hunk 7 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/PDBWriter.cpp]: Or moving the `MaeWriter` class to a separate header, that should work too.

**Review by greglandrum — APPROVED** [review]
LGTM
